Compiler can generate delegate with direct pointer when referencing static members

MoFtZ opened the pull request:
This PR attempts to fix #8295.

Apologies in advance, I'm not an F# programmer, so there's probably a much better way to implement what I'm trying to do.

It looks like a big diff, but most of it are changes to the indentation.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=9017) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=9017) before we can accept your contribution.<br/><br/>:x: MoFtZ [sign now](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=9017)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/fsharp?pullRequest=9017) it.</sub>

smoothdeveloper commented:
@MoFtZ thanks! I think one thing which will be needed is to add a specific test if none of the IL test fail.

If such existing tests are failing, it means this pattern of a delegate pointing to static method is already exhibited under those tests, and the would need adjustment of baseline files along those lines: https://github.com/dotnet/fsharp/blob/master/TESTGUIDE.md#baselines

smoothdeveloper commented on src/fsharp/IlxGen.fs:
```diff
@@ -4781,54 +4781,87 @@ and GenDelegateExpr cenv cgbuf eenvouter expr (TObjExprMethod((TSlotSig(_, deleg
+                | _ -> None
+            | _ -> None
+
+    let stripBodyExpr =
```
probably can be written as

```fsharp
let stripBodyExpr body =
  match stripExpr body with
  | Expr.Let (_, letBody, _, _) -> letBody
  | _ -> body
```

same remark for above function, you can declare function with parameters as a let binding.

MoFtZ replied:
Ooo... thanks =)

smoothdeveloper commented on src/fsharp/IlxGen.fs:
```diff
@@ -4781,54 +4781,87 @@ and GenDelegateExpr cenv cgbuf eenvouter expr (TObjExprMethod((TSlotSig(_, deleg
+        fun body ->
+            match body with
+            | Expr.App (f, _, _, _, m) ->
+                match f with
```
The two top matches can possibly be combined if you feel it remains readable enough, it will save one nesting level.

```fsharp
match body with
| Expr.App (Expr.Val (vref, _, _), _, _, _, m) ->
 ...
```

MoFtZ replied:
Ah, nice - yes, I remember seeing something similar in the existing code.

MoFtZ commented:
Thanks @smoothdeveloper, I will look into those baseline tests.

MoFtZ commented:
OK, I've added some unit tests for existing behavior, and proposed behavior. And I've fixed up the code to deal with some missed conditions.

However, I've run into an interesting problem with the optimizer. When optimizations are turned on (default), the `lambdaInlineThreshold` configuration setting is set to 6. This causes the delegate to be inlined before the new code gets a chance to to inspect it. Does anyone have any suggestions for how to deal with this scenario?

KevinRansom commented:
@MoFtZ , what is happening is that the optimizer is inlineing the code as far up the tree as it can., which because your tests have a do nothing static method, results in us generating an Invoke method which also does nothing.

If you have a static method that is not in-line-able, because of size ... your code does exactly the right thing ....

Eg.
I just added a big ol' body for C.M() -- The window on the right is code generated by the rtm compiler, on the right is with your enhancement.
![image](https://user-images.githubusercontent.com/5175830/80427648-157d5600-889d-11ea-9e7e-933a267e6241.png)

You should look things over, but I think there is nothing more you need to do, beyond, adding  test-cases with non-inline-able bodies ...

Did what I just say make any sense at all?  It seems to in my head :-)

MoFtZ commented:
Thanks for looking into this @KevinRansom. Yes, what you said makes perfect sense.

My concerns are related to the [original issue](#8295), which was raised because of issues with a C# library that could only deal with static methods. This means that F# developers trying to use that C# library will not be able to port any sample code that has a short static method. This is not ideal, but also not a deal-breaker.

For real world usage, this PR should work just fine, and I'm happy for it to be merged.

A final minor nitpick, it would have been nice to be able to deal with empty static members by calling the original static member, rather than creating a closure class with an empty Invoke - no need to allocate memory for an object, and no need to emit a closure class.

KevinRansom commented:
@MoFtZ, special casing empty methods, is probably not really worthwhile, since, Invoke() can contain the in-lined body of CM, for however, much code we allow on inlining.

MoFtZ commented:
How should we proceed? Would you like me to rewrite the unit tests with a larger method body?

KevinRansom commented:
Yes, this change is good, and has a nice benefit.  But I do want to think a bit about the actual issue you were addressing, because it does seem that when we generate the 
````
     return new Action(new CreateAction@5().Invoke);
````

And the Invoke implementation, there is no real reason not to make it static when nothing is captured by the closure.  It would additionally save us a pointless allocation.

MoFtZ commented:
In the case when there is nothing to capture, could we emit a static class with a static Invoke method? The delegate could then call into this static Invoke method directly. This would work regardless of the inline optimization.

KevinRansom commented:
@moftz, yes that is what I was thinking of looking at.  It would be super helpful, if you could give it a shot,

MoFtZ commented:
OK, I've reworked all the commits. The test cases now run with optimizations turned on, and my initial changes for detecting/extracting a static method have been abandoned.

The new changes follow the original code, with a new check for captured variables. If there are no captured variables, it will generate a static Invoke method, and a static class to contain it.

I'm not particularly happy with how I have generated the static class (it strips the empty constructor, and makes a lot of assumptions), but I was not comfortable changing the innards of `EraseClosures.convIlxClosureDef`.

TIHan commented:
@MoFtZ - When I get a chance, I'm going to take a look at this.

MoFtZ commented:
This PR alters the code generation path for `GenDelegateExpr`. AFAIK, this is the only location that emits IL for delegates.

My thoughts on moving this logic into the `Optimization` phase would be:
1. Alter the `Optimization` phase to identify non-capturing delegates and emit a new TAST node/expression
1. Alter the `Code Generation` phase to emit suitable IL e.g. `GenNonCapturingDelegateExpr`

This would move the "non-capturing" detection logic into the `Optimization` phase, and introduce a new/separate code path for emitting IL.

If this sounds reasonable, I can alter the PR to match.

Alternatively, does anyone have suggestions for how to move this logic into the `Optimization` phase?

TIHan commented:
@MoFtZ if we wanted to move it out of IlxGen, your approach would make sense.

I think doing the work in IlxGen is fine. I just wanted to know what the other alternatives were. Trying to build a new TAST node for this would be much more work; I'm not sure what the benefits would be.

MoFtZ commented:
Thanks for the feedback @TIHan.

Are you able to retrigger the CI/CD pipeline? Error log say:
```
##[error]An error occurred while provisioning resources (Error Type: Disconnect).
,##[warning]Received request to deprovision: The request was cancelled by the remote provider.
```

KevinRansom commented:
Thank you for this contribution.
